File: tests/test_audit_regressions.mjs
/* SO CAI - cac phat hien cua ban ra soat 2026-08-30, viet thanh test.
   Xem docs/AUDIT-2026-08-30.md de biet chi tiet tung muc.

   BO TEST NAY DO CO CHU Y. Moi muc con OPEN la mot loi CHUA sua. Sua xong thi
   no tu chuyen sang FIXED. Khi tat ca FIXED, file nay xanh va tro thanh test
   chong tai phat - dung xoa no di sau khi sua.

   KHONG can mang: tat ca deu chay tren ham thuan hoac tren van ban nguon.

   Chay:  node tests/test_audit_regressions.mjs                              */
import { boot, bodyMarkup, appJs } from "./harness.mjs";

/* Ten field cua tram Sai Gon. Dung de kiem tra cac cho con gan cung theo SO
   FIELD: Sai Gon dat nhiet do o field3, Bao Loc dat o field1. */
const SAIGON_FIELDS = {
  name: "MakerLab Station",
  field1: "Wind speed", field2: "Wind direction", field3: "Temperature",
  field4: "Pressure", field5: "Light", field6: "Humidity",
  field7: "Noise", field8: "PM2.5",
};

const findings = [];
const finding = (id, title, probe) => findings.push({ id, title, probe });

/* -------------------------------------------------------------------------
   A1 - Trang bia ten/don vi/dai hop ly cho field mot kenh la KHONG he co.
   ------------------------------------------------------------------------- */
finding("A1", "Kenh la khong duoc bia ten + don vi cho field no khong co", () => {
  const { run } = boot(3448221, { offline: true });
  run(`applyChannelMetadata({name:"Solar Inverter Log",field1:"Voltage",field2:"Current",field3:"Battery"})`);
  const invented = JSON.parse(run(
    `JSON.stringify(Object.entries(F).filter(([k])=>['field4','field5','field6','field7','field8'].includes(k))
       .filter(([,c])=>c.unit||c.range).map(([k,c])=>k+'='+c.name+' '+(c.unit||'')))`,
  ));
  return {
    open: invented.length > 0,
    evidence: invented.length
      ? `kenh chi khai bao 3 field, trang van gan: ${invented.join(" | ")}`
      : "field khong khai bao -> khong ten, khong don vi, khong dai",
  };
});

/* -------------------------------------------------------------------------
   A2 - ZERO_SIGNATURE con gan cung field1/2/3, tuc la doc SAI o Sai Gon.
   ------------------------------------------------------------------------- */
finding("A2", "Dau hieu 'cam bien vua khoi dong lai' phai theo VAI TRO, khong theo so field", () => {
  const { run } = boot(3448221, { offline: true });
  run(`applyChannelMetadata(${JSON.stringify(SAIGON_FIELDS)})`);

  /* Tren Sai Gon, so 0 khi cam bien chua am la o nhiet do (field3),
     do am (field6) va anh sang (field5). */
  const realWarmup = run(`bootReason({status:'ALL_OK',field3:0,field6:0,field5:0})`);
  /* Con field1/field2/field3 = 0 tren Sai Gon la: lang gio, huong 0, 0 do C.
     Lang gio la chuyen binh thuong hang dem - khong phai cam bien reboot. */
  const calmNight = run(`bootReason({status:'ALL_OK',field1:0,field2:0,field3:0})`);

  const misses = realWarmup !== "sensor";
  const falsePositive = calmNight === "sensor";
  return {
    open: misses || falsePositive,
    evidence: [
      misses ? "BO SOT: warm-up that (temp/hum/light = 0) -> " + JSON.stringify(realWarmup) : null,
      falsePositive ? "BAO NHAM: dem lang gio (wind/dir/temp = 0) -> 'sensor'" : null,
    ].filter(Boolean).join("; ") || "theo vai tro, dung ca hai chieu",
  };
});

/* -------------------------------------------------------------------------
   A3 - Chu tren trang gan cung con so 8, trong khi trang nhan MOI kenh.
   ------------------------------------------------------------------------- */
finding("A3", "Chu hien tren trang khong duoc gan cung 'eight' khi kenh la co the it field hon", () => {
  const inMarkup = (bodyMarkup.match(/\beight\b/gi) || []).length;
  /* Chuoi verdict nam trong JS, khong nam trong markup, nen dem rieng. */
  const inVerdict = (appJs.match(/all eight sensors/gi) || []).length;
  return {
    open: inMarkup + inVerdict > 0,
    evidence: inMarkup + inVerdict
      ? `markup: ${inMarkup} cho, chuoi verdict: ${inVerdict} cho`
      : "so field lay tu chinh kenh",
  };
});

/* -------------------------------------------------------------------------
   A4 - Ghi chu moc du lieu ke rieng cau chuyen cua Sai Gon cho MOI kenh.
   ------------------------------------------------------------------------- */
finding("A4", "Ghi chu 'data epoch' khong duoc ke ly do cua rieng mot tram", () => {
  const hard = /station reinstall and firmware update|different wind scaling/.test(appJs);
  return {
    open: hard,
    evidence: hard
      ? "chuoi epochNote gan cung 'station reinstall'/'different wind scaling' - "
        + "sai ngay khi Bao Loc duoc chot moc (ly do la field map khac, khong phai thang do gio)"
      : "ly do moc lay theo tung kenh",
  };
});

/* -------------------------------------------------------------------------
   A5 - Lan ve dau tien hien ten cua kenh TRUOC do, chong len id kenh moi.
   ------------------------------------------------------------------------- */
finding("A5", "Lan ve dau tien khong duoc dan ten tram nay len id cua tram kia", () => {
  const { document, el } = boot(3428136, { offline: true });
  const title = document.title;
  const shownName = el("#channelName").textContent;
  const wrong = /Bao Loc/i.test(title) || /Bao Loc/i.test(shownName);
  return {
    open: wrong,
    evidence: wrong
      ? `?channel=3428136 nhung title="${title}", #channelName="${shownName}"`
      : `title="${title}"`,
  };
});

/* ------------------------------- chay ------------------------------------ */
let open = 0;
console.log("SO CAI DASHBOARD - ban ra soat 2026-08-30\n");
for (const f of findings) {
  let r;
  try {
    r = f.probe();
  } catch (error) {
    r = { open: true, evidence: "probe nem loi: " + (error && error.message) };
  }
  if (r.open) open++;
  console.log(`  [${r.open ? "OPEN " : "FIXED"}] ${f.id}  ${f.title}`);
  console.log(`           ${r.evidence}\n`);
}
console.log(`${findings.length - open} FIXED, ${open} OPEN`);
if (open) console.log("\nBo test nay do la DUNG cho toi khi cac muc tren duoc sua.");
process.exit(open ? 1 : 0);

File: tests/test_audit_regressions.py
"""SO CAI - cac phat hien cua ban ra soat 2026-08-30 ve FIRMWARE.

Xem docs/AUDIT-2026-08-30.md de biet chi tiet tung muc.

BO TEST NAY DO CO CHU Y. Moi muc con OPEN la mot loi CHUA sua. Sua xong thi no
tu chuyen sang FIXED. Khi tat ca FIXED, file nay xanh va tro thanh test chong
tai phat - dung xoa no di sau khi sua.

Khong can board, khong can mang.

Chay:  py -3 tests/test_audit_regressions.py
"""

import pathlib
import re
import sys
import types
import time as _time

ROOT = pathlib.Path(__file__).resolve().parent.parent
FW = ROOT / "firmware"
DOCS = ROOT / "docs"

FINDINGS = []


def finding(fid, title):
    def wrap(fn):
        FINDINGS.append((fid, title, fn))
        return fn
    return wrap


# ---------------------------------------------------------------------------
# Nap firmware tren PC: stub cac module chi co tren board, cat truoc vong lap.
# ---------------------------------------------------------------------------
for _attr, _fn in (
    ("sleep_ms", lambda ms: None),
    ("sleep_us", lambda us: None),
    ("ticks_ms", lambda: 0),
    ("ticks_add", lambda t, d: t + d),
    ("ticks_diff", lambda a, b: a - b),
):
    setattr(_time, _attr, _fn)


def _stub_board():
    machine = types.ModuleType("machine")
    machine.WDT = lambda *a, **k: None
    machine.WDT_RESET, machine.SOFT_RESET = 3, 5
    machine.reset_cause = lambda: 1
    machine.reset = lambda: None
    machine.UART = lambda *a, **k: types.SimpleNamespace(
        any=lambda: 0, read=lambda *_: b"", write=lambda *_: None)
    machine.Pin = machine.I2C = machine.SoftI2C = lambda *a, **k: None

    network = types.ModuleType("network")
    network.STA_IF = 0
    network.WLAN = lambda *a, **k: types.SimpleNamespace(
        active=lambda *_: None, isconnected=lambda: False, connect=lambda *_: None,
        disconnect=lambda: None, ifconfig=lambda: ("0",) * 4, status=lambda: 0)

    requests = types.ModuleType("requests")
    requests.post = lambda *a, **k: None

    ntptime = types.ModuleType("ntptime")
    ntptime.timeout = 1
    ntptime.settime = lambda: None

    sys.modules.update(machine=machine, network=network,
                       requests=requests, ntptime=ntptime)


def load(name):
    """Nap phan module cua mot firmware (truoc '# MAIN PROGRAM')."""
    _stub_board()
    path = FW / name
    text = path.read_text(encoding="utf-8")
    head = text[: text.index("# MAIN PROGRAM")].rsplit("# " + "=" * 57, 1)[0]
    mod = types.ModuleType("fw_" + name)
    # Firmware in canh bao ra stdout khi thieu station_secrets.py - dung y,
    # nhung o day chi lam ban ket qua, nen nuot lai.
    buf, sys.stdout = sys.stdout, open(__import__("os").devnull, "w", encoding="utf-8")
    try:
        exec(compile(head, str(path), "exec"), mod.__dict__)
    finally:
        sys.stdout.close()
        sys.stdout = buf
    return mod


def source(name):
    return (FW / name).read_text(encoding="utf-8")


# Gia tri toi da mot khung Modbus DUNG CRC van co the mang: thanh ghi u16 day.
# Firmware chia 10 cho ap suat / tieng on / toc do gio, nen 65535/10 = 6553.5.
GLITCH = {
    "temperature": 25.0, "humidity": 50.0, "noise": 6553.5,
    "pm2_5": 65535, "pm10": 65535, "pressure": 6553.5,
    "light": 4294967295, "wind_speed": 6553.5, "wind_angle": 65535,
}


def lcd_rows(fw, name, snap, link):
    if name == "saigon.py":
        return fw.lcd_lines(snap["temperature"], snap["pm2_5"], snap["light"],
                            snap["wind_speed"], snap["wind_angle"], link)
    return fw.lcd_lines(snap, link)


@finding("B1", "LCD khong duoc cat giua mot con so, va khong duoc nuot ky tu bao trang thai")
def b1():
    broken = []
    for name in ("bao_loc.py", "saigon.py"):
        fw = load(name)
        for link in (fw.LINK_OK, fw.LINK_NO_WIFI):
            rows = lcd_rows(fw, name, GLITCH, link)
            if not rows[-1].endswith(link):
                broken.append("%s: mat ky tu %r o dong cuoi -> |%s|"
                              % (name, link, rows[-1]))
            for row in rows:
                if len(row) > fw.LCD_COLS:
                    broken.append("%s: tran cot -> |%s| (%d)" % (name, row, len(row)))
    return broken, ("; ".join(broken[:3]) if broken
                    else "gia tri u16 day van giu duoc bo cuc va ky tu trang thai")


@finding("B2", "Chuoi hien tren LCD khong duoc gan cung Channel ID")
def b2():
    hits = []
    for name in ("bao_loc.py", "saigon.py"):
        text = source(name)
        match = re.search(r"^CHANNEL_ID\s*=\s*(\d+)", text, re.M)
        if not match:
            hits.append("%s: khong tim thay CHANNEL_ID" % name)
            continue
        cid = match.group(1)
        for lit in re.findall(r'"[^"\n]*"', text):
            if cid in lit:
                hits.append("%s: %s" % (name, lit))
    return hits, ("; ".join(hits) if hits
                  else "moi cho hien channel deu lay tu CHANNEL_ID")


@finding("B3", "Hai firmware phai co cung cong tac debug tai cho (CLOUD_UPLOAD_ENABLED)")
def b3():
    missing = []
    for name in ("bao_loc.py", "saigon.py"):
        text = source(name)
        declared = re.search(r"^CLOUD_UPLOAD_ENABLED\s*=", text, re.M) is not None
        mentioned = "CLOUD_UPLOAD_ENABLED" in text
        if mentioned and not declared:
            missing.append("%s: nhac CLOUD_UPLOAD_ENABLED nhung KHONG khai bao "
                           "-> LINK_OFF la code chet, khong tat duoc upload de debug" % name)
        elif not declared:
            missing.append("%s: khong co CLOUD_UPLOAD_ENABLED" % name)
    return missing, ("; ".join(missing) if missing
                     else "ca hai firmware deu tat duoc upload de debug tai tram")


@finding("B4", "Khong con tro toi ten file cu 'siuuuu.py'")
def b4():
    # Chi quet noi RA LENH cho nguoi dung: firmware va README. `docs/` la bien
    # ban lich su - postmortem 2026-07-22 va ban ra soat deu PHAI nhac ten file
    # cu de ke lai chuyen da xay ra, do la ban ghi dung chu khong phai tham
    # chieu chet. Bat ca docs se bien test nay thanh thu phat nguoi ghi su.
    stale = []
    for path in sorted(FW.glob("*.py")) + [ROOT / "README.md"]:
        if path.name == "station_secrets.py":       # gitignore, khong thuoc repo
            continue
        for i, line in enumerate(path.read_text(encoding="utf-8").splitlines(), 1):
            if "siuuuu" in line:
                stale.append("%s:%d" % (path.relative_to(ROOT).as_posix(), i))
    return stale, ("; ".join(stale) if stale
                   else "khong con huong dan nao tro toi ten file da doi")


@finding("B5", "Cau 'DO TRAN BOARD THAT' phai noi ro do tren BOARD NAO")
def b5():
    post = (DOCS / "ESP32_HTTPS_TLS_POSTMORTEM_2026-07-22.md").read_text(encoding="utf-8")
    # Postmortem ghi runtime la ESP32_GENERIC va file la siuuuu.py -> tram Sai Gon
    # (ESP32-WROOM). Bao Loc chay ESP32-S3, mot con chip khac.
    measured_on_wroom = "ESP32_GENERIC" in post
    bad = []
    for name in ("bao_loc.py", "saigon.py"):
        text = source(name)
        for i, line in enumerate(text.splitlines(), 1):
            if "BOARD TH" in line.upper() and "2026-07-22" in line:
                window = "\n".join(text.splitlines()[i - 1:i + 8])
                names_board = re.search(r"WROOM|ESP32_GENERIC|Sai Gon|Sài Gòn", window)
                if not names_board:
                    bad.append("%s:%d neu 'do tren board that' ma khong noi board nao" % (name, i))
    if bad and not measured_on_wroom:
        bad.append("(khong xac minh duoc board tu postmortem)")
    return bad, ("; ".join(bad) if bad
                 else "moi trich dan phep do deu ghi ro board da do")


def main():
    print("SO CAI FIRMWARE - ban ra soat 2026-08-30\n")
    open_count = 0
    for fid, title, fn in FINDINGS:
        try:
            problems, evidence = fn()
        except Exception as error:                  # noqa: BLE001 - probe hong = con OPEN
            problems, evidence = [1], "probe nem loi: %r" % (error,)
        is_open = bool(problems)
        open_count += is_open
        print("  [%s] %s  %s" % ("OPEN " if is_open else "FIXED", fid, title))
        print("           %s\n" % evidence)
    print("%d FIXED, %d OPEN" % (len(FINDINGS) - open_count, open_count))
    if open_count:
        print("\nBo test nay do la DUNG cho toi khi cac muc tren duoc sua.")
    return 1 if open_count else 0


if __name__ == "__main__":
    sys.exit(main())
